HPCC-19489 STD.File.RenameLogicalFile should support OVERWRITE

AttilaVamos opened the pull request:
Add allowOverwrite parameter to STD.File.RenameLogicalFile

Add code if overwrite is allowed to:
- check source and target file existence
- remove existing target file

Extend fileservice.ecl to test allowOverwrite option

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [x] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [ ] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [x] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->
Changes are tested manually.
<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

AttilaVamos commented:
@jakesmith please review

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 32715cb2b9b564b6997227c0ac78ed9878ac06e8
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |89 |89 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   20 sec (00:00:20) |  173 sec (00:02:53) |   93 sec (00:01:33) |   21 sec (00:00:21) |   43 sec (00:00:43) | 1359 sec (00:22:39) |   20 sec (00:00:20) | 1729 sec (00:28:49) |

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -523,31 +523,63 @@ FILESERVICES_API void FILESERVICES_CALL fsSetReadOnly(ICodeContext *ctx, const c
-FILESERVICES_API void FILESERVICES_CALL fsRenameLogicalFile(ICodeContext *ctx, const char *oldname, const char *newname)
+FILESERVICES_API void FILESERVICES_CALL implementRenameLogicalFile(ICodeContext *ctx, const char *oldname, const char *newname, const bool overwrite)
 {
     StringBuffer lfn, nlfn;
```
pedantic/not new: but would be slightly clearer if lfn was olfn, or these were oldLfn/newLfn or similar.

AttilaVamos replied:
Changed

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -523,31 +523,63 @@ FILESERVICES_API void FILESERVICES_CALL fsSetReadOnly(ICodeContext *ctx, const c
+    if (overwrite)
+    {
+        if ( distributedDirectory.exists(nlfn.str(), ctx->queryUserDescriptor(), false, false) &&
+             distributedDirectory.exists(lfn.str(), ctx->queryUserDescriptor(), false, false))
```
If old lfn doesn't exist, there's no point in continuing at all, with or without overwrite. (renamePhysical will fail).
So would be clearer if checked if old existed and error out before anything else.

AttilaVamos replied:
Changed

jakesmith commented on plugins/fileservices/fileservices.cpp:
```diff
@@ -523,31 +523,63 @@ FILESERVICES_API void FILESERVICES_CALL fsSetReadOnly(ICodeContext *ctx, const c
+                    s.append("') done");
+                WUmessage(ctx, SeverityInformation, NULL, s.str());
+                AuditMessage(ctx, "DeleteLogicalFile", nlfn.str());
+            }
```
lines 541 to 552 are same as code in fsDeleteLogicalFile.
should be commoned up, i.e. both call common function, or possibly for this just to calll fsDeleteLogicalFile directly.

AttilaVamos replied:
Changed

jakesmith commented on testing/regress/ecl/fileservice.ecl:
```diff
@@ -33,5 +33,10 @@ SEQUENTIAL(
+  OUTPUT(ds, , prefix + 'renametest.d00', OVERWRITE),
+  OUTPUT(ds, , prefix + 'afterrename1.d00', OVERWRITE),
+  // Remane with overwrite allowed
+  File.RenameLogicalFile(prefix + 'renametest.d00', prefix + 'afterrename1.d00', true),
```
I think there may be some untested problems.
1) Rename external or foreign with overwrite.
Both are disallowed and will error on rename, but the delete will have already happened.. 
Foreign delete should fail, so may be ok.
External delete .. I'm not sure, worth testing.
i.e. need to make sure it doesn't succeed to delete - but then fail on rename.
2) Rename in a transaction.
Actually fileservices.ecl doesn't have any transaction tests at the moment.
I think there may be problems with rename in particular, related to the way delete happens.
See addDelayedDelete/CDelayedDelete in dadfs.cpp

AttilaVamos replied:
As we discussed, I will test this delete, rename in transaction.

jakesmith commented on testing/regress/ecl/fileservice.ecl:
```diff
@@ -33,5 +33,10 @@ SEQUENTIAL(
+  File.DeleteLogicalFile(prefix + 'scope1::afterrename3.d00'),
+  OUTPUT(ds, , prefix + 'renametest.d00', OVERWRITE),
+  OUTPUT(ds, , prefix + 'afterrename1.d00', OVERWRITE),
+  // Remane with overwrite allowed
```
typo: Rename

AttilaVamos replied:
Changed

jakesmith commented on testing/regress/ecl/fileservice.ecl:
```diff
@@ -33,5 +33,10 @@ SEQUENTIAL(
+  OUTPUT(ds, , prefix + 'afterrename1.d00', OVERWRITE),
+  // Remane with overwrite allowed
+  File.RenameLogicalFile(prefix + 'renametest.d00', prefix + 'afterrename1.d00', true),
+  File.DeleteLogicalFile(prefix + 'afterrename1.d00'),
```
Would be good to read the renamed file after the test to verify before deleting it.

AttilaVamos replied:
Added

jakesmith commented:
@AttilaVamos - added some comments.

AttilaVamos commented:
@jakesmith updated, please re-review.

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 6581d6d8300e918942c6eab23ba47193d1362787
Build: success
Install hpccsystems-platform-community_7.0.0-rc1.el7.x86_64.rpm
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |94 |94 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |805 |805 |0 |
|test (thor) |729 |729 |0 |
|test (roxie) |878 |878 |0 |


HPCC Stop: OK
HPCC Uninstall: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   16 sec (00:00:16) |  173 sec (00:02:53) |   93 sec (00:01:33) |   20 sec (00:00:20) |   42 sec (00:00:42) | 1365 sec (00:22:45) |   19 sec (00:00:19) | 1728 sec (00:28:48) |

jakesmith commented:
@AttilaVamos - looks good, but 1 question.

Did the "As we discussed, I will test this delete, rename in transaction."  happen, is there a separate JIRA?

richardkchapman commented:
@AttilaVamos Please squash

AttilaVamos commented:
@jakesmith I will open a JIRA or perhaps two to cover rename logical file in transaction, for foreign and external file.

jakesmith commented:
@AttilaVamos - please squash ready for merge and reference the new JIRA's in this issues JIRA.

AttilaVamos committed:
HPCC-19489 STD.File.RenameLogicalFile should support OVERWRITE

Add allowOverwrite parameter to STD.File.RenameLogicalFile

Add code if overwrite is allowed to:
- check source and target file existence
- remove existing target file

Extend fileservice.ecl to test allowOverwrite option

Signed-off-by: Attila Vamos <attila.vamos@gmail.com>
